### ISSUE TITLE
Difference in behavior between .NET and Microsoft schemas with configuration merging

### ISSUE THREAD

**Issue by CuddleBunny** [microsoft/FeatureManagement-Dotnet#507]
This isn't necessarily an issue, but it should probably be called out somewhere: when switching from the .NET schema to the Microsoft schema you lose the ability to spread definitions across multiple configuration files (ie. appsettings.json and appsettings.Prod.json) because `feature_management:feature_flags` is an array which causes additional configuration sources to overwrite the original. This isn't the case with the .NET schema because the collection is an object and merges fine.

**Comment by rossgrambo** [comment]
Hey @CuddleBunny, thank you for calling this out and following up on it.
 
It's unfortunate that this change to unify our libraries caused this to work differently. 

The issue with Configuration merging is discussed [here](https://github.com/dotnet/runtime/issues/36569). Sadly, it's upstream of FeatureManagement- as this library reads whatever is made available by IConfiguration. 

Theres two workarounds I'm aware of:
### Define your own `IFeatureDefinitionProvider`

I know this could solve the issue if you manually read the files and merge them how you see appropriate. Our default implementation reads from `IConfiguration` but you don't have to in your version.

It might be possible to have your implementation call each `JsonConfigurationProvider` within the `Configuration` separately so you can handle the merge yourself and don't have to re-read these files. If this is possible and not too heavy it could be something we consider adding to the default `IFeatureDefintionProvider`. But I think that's unlikely because there's not one way to merge that works for everyone.

### `:#` syntax (Not recommended)
Using `:#` to override arrays appropriately. 
  When defining an array in json to be used in IConfiguration, you can define it like:
  
  ```json
  "feature_management": {
    "feature_flags:0": {
      "id": "OnTestFeature",
      "enabled": true
    }
  }
  ```

In your example, you could define your `appsettings.prod.json` as
```json
{
  "feature_management": {
    "feature_flags:2": {
      "id": "ProdOnlyThing",
      "enabled": true
    }
  }
}
```

Unfortunately, this is finicky to line things up when you're trying to override correctly. 

If your only goal is to merge I believe you start at a large number (like `:1000`) in all of your prod flags and they would always append to the end of the other flags. This is not an intended feature so there's no guarantee that it'll continue to work but it seems to work for me at first glance.

**Comment by CuddleBunny** [comment]
I've considered the IFeatureDefinitionProvider route already and will likely take that path. We should close this issue, but it could probably be called out in the docs somewhere as a possible gotcha, especially because the default dotnet templates have configuration per environment built in so it's widely used.

**Comment by pinkfloydx33** [comment]
@CuddleBunny   did you ever come up with something reasonable?

**Comment by CuddleBunny** [comment]
> [@CuddleBunny](https://github.com/CuddleBunny) did you ever come up with something reasonable?

For the moment I've removed all feature flags from appsettings.json and have the full list duplicated in every environment specific configuration. I haven't had the bandwidth to write the custom `IFeatureDefinitionProvider` but when I do I'll be sure to post an example here.

**Comment by jimmyca15** [comment]
@rossgrambo provided a solution using different offsets in the feature flag array across different appsettings.json file. He mentioned _not recommended_, but I actually think this is a reasonable approach. If you define the flags in your appsettings.json with some high offset, which is different in each file, there should be no collisions. This saves the effort of having to create a custom IFeatureDefinitionProvider.

This is a bit of a battle with the .NET configuration system which doesn't allow different configuration providers (e.g. two different json files, or a json file + app configuration) to know about each other and adjust accordingly.

**Comment by pinkfloydx33** [comment]
I fortunately only need this for development setups where I provide the settings packaged in a docker image and devs might override them. No azure app config involved.

Based on the suggestion above, I actually discovered that due to the way this library reads the configuration (it just walks the children directly) , that you can use named keys instead of integers for the feature flags:

```
{
   "feature_management": {
       "feature_flags:named": {

      }
  }
}
```

This is much easier to reason about and remember, if the goal is layering. I wouldn't do this in production but like I said my target is local development scenarios so this works fine.

Unfortunately it stops there, as the feature definition provider doesn't work quite the same for variants. In that case it tries to parse the key as an integer and if it fails it will skip the variant section.

**Comment by jimmyca15** [comment]
Interesting, I am thinking about the difference in support there.

**Comment by pinkfloydx33** [comment]
Yeah I accidentally discovered that, but then dug into the source a little. 

It totally falls apart if anything is coming from Azure AppConfig since the provider converts the data into arrays (and thus numbered keys) and there's no way to know what order the features or variants are going to come from AAC. Use of arrays is unfortunate with the way the dotnet configuration provider performs configuration layering

**Comment by abatishchev** [comment]
I'm adopting FM in my ASP.NET Core app and this issue worries me. I basically have 2 requireements:

1. Define features in the root config (appSettings.json)
2. Be able to turn them on and off (i.e. override) based in sub-configs (cloud, environment, region)

Would I be able to achieve the latter using the default format which is an array?

For the named keys "hack", I found that `id` is still mandatory, i.e.:

```json
"feature_management": {
  "feature_flags:FooBar": {
    "id": "FooBar",
    "enabled": true,
    "telemetry": {
      "enabled": true
    }
  }
}
```

Otherwise, it fails to bind and `GetFeatureDefinitionAsync` returns nothing.

**Comment by jimmyca15** [comment]
@abatishchev so you are only using json files for configuration?

**Comment by abatishchev** [comment]
@jimmyca15 yes. the underlying hosting platform is Service Fabric. But it's running my app as a guest executable (simply starts it up and makes sure it doesn't go down) so the app's code doesn't have any dependencies on SF, uses the standard ASP.NET tooling: ILogger, IOptions, DI, etc.

**Comment by abatishchev** [comment]
Is it worth starting a new issue to outline the problem and discuss options how to address it?

>one would have to duplicate the entire flag definition across json files in order to override

Isn't it an option already today when one would duplicate all entries in the array but with slightly different values (hence, overriding them at a lower level)?

I also would like to mention that the best solution would be the one which doesn't require duplication. Does this one:

```
"feature_management": {
  "feature_flags:FooBar": {
    "id": "FooBar",
    "enabled": true,
    "telemetry": {
      "enabled": true
    }
  }
}
```json

require duplication? I think (hope) it does not.

**Comment by jimmyca15** [comment]
@abatishchev 

With this solution, if you define the flag across multiple files/configuration providers we would use the last flag definition and ignore all others.

**Comment by pinkfloydx33** [comment]
Sorry I missed it. 

We distribute the app settings "base" in a docker image used within a docker compose enviroment. The users in this instance are primarily front end developers making use of a full stack enviroment. They don't care about the schema of the feature flags, they just want to toggle them on and off for testing/development. This works well with the dotnet schema since it's literally a true/false value and you don't need them to learn anything more. For non-variants this is still viable. 

With variant features, it's a similar issue. Either they need to toggle a feature on/off _or_ modify some of the variant configuration data (the arbitrary, set-by-us data) without caring about allocations (or what that even means).

In the old style, layering the variant was still a little clunky but at least it could be minimized if the bulk of it was in the base settings file. Now I need them to understand how to override every piece of it and what it means, when they truly don't care nor have a reason to care.

---

NB. We use AppConfiguration in deployed enviroments. There's no layering concern there. You get the *whole* feature flag/variant or you get _nothing_. So that's cool. My problem is just for my local development setups... But not everyone uses AppConfig. As highlighted by this thread folks distribute the flags as they do with any other configuration and for them the layering story has (IMO) gotten worse and no longer matches the rest of the app settings they are distributing and layering.

**Comment by zhiyuanliang-ms** [comment]
> We distribute the app settings "base" in a docker image used within a docker compose enviroment. The users in this instance are primarily front end developers making use of a full stack enviroment. They don't care about the schema of the feature flags, they just want to toggle them on and off for testing/development. This works well with the dotnet schema since it's literally a true/false value and you don't need them to learn anything more. For non-variants this is still viable.

Hi, @pinkfloydx33 For your scenario, I think the feature flag layering behavior you would expect only needs to happen on the `enabled` field. Right?

**Comment by zhiyuanliang-ms** [comment]
@pinkfloydx33 @abatishchev @CuddleBunny @Quentin-IMA

We do not currently plan to revert the feature flags array "concatenation" behavior introduced in 4.2.1.


I think there could be two options here:

1. We add a flag to disable the "concatenation" behavior, which means it will fallback to the behavior before v4.2. In this case, people can still use the "index" work-around. ([named index](https://github.com/microsoft/FeatureManagement-Dotnet/issues/507#issuecomment-2831015437) or [number index](https://github.com/microsoft/FeatureManagement-Dotnet/issues/550#issue-3240033898)).

2. We introduce a new feature flag merge bahavior. I have a draft POC [PR](https://github.com/microsoft/FeatureManagement-Dotnet/pull/551) for it. 
The feature manager will merge all feature flag declarations with the same `id` (from all configuration sources). 

e.g.

```json
"feature_flags": [
    {
        "id": "FeatureA",
        "enabled": false，
    },
    {
        "id": "FeatureA",
        "conditions": {
            "client_filters": [...]
        }
    },
    {
        "id": "FeatureA",
        "enabled": true
    }
]
```
The final feature definition will be
```json
{
    "id": "FeatureA",
    "enabled": true,
    "conditions": {
        "client_filters": [...]
    }
}
```

We can add a flag to switch between the array "concatenation" behavior and the feature flag "merge" behavior.

Which option do you prefer?

**Comment by jimmyca15** [comment]
My vote is for option 1 (add a flag to disable the "concatenation" behavior) where we just let the .NET provider configuration system behavior win out vs defining our own semantics.

**Comment by zhiyuanliang-ms** [comment]
Hey, @pinkfloydx33 @abatishchev @CuddleBunny @Quentin-IMA

I sent out the PR #552. In the PR, a new `ConfigurationFeatureDefinitionProviderOptions` is added.

```c#
/// <summary>
/// Options that control the behavior of the <see cref="ConfigurationFeatureDefinitionProvider"/>.
/// </summary>
/// <remarks>
/// The <see cref="ConfigurationFeatureDefinitionProvider"/> uses custom configuration merging logic 
/// for Microsoft schema feature flags to ensure that feature flags with the same ID from different 
/// configuration sources are merged correctly based on their logical identity rather than array position.
/// By default, the provider bypasses .NET's native array merging behavior which merges arrays by index 
/// position and can lead to unexpected results when feature flags are defined across multiple configuration sources.
///
/// Consider the following configuration sources:
/// Configuration Source 1:
/// {
///   "feature_management": {
///     "feature_flags": [
///       {
///         "id": "feature1",
///         "enabled": true
///       },
///       {
///         "id": "feature2", 
///         "enabled": false
///       }
///     ]
///   }
/// }
/// 
/// Configuration Source 2:
/// {
///   "feature_management": {
///     "feature_flags": [
///       {
///         "id": "feature2",
///         "enabled": true
///       }
///     ]
///   }
/// }
/// 
/// With custom merging (default behavior):
/// - feature1: enabled = true
/// - feature2: enabled = true (last declaration wins)
/// 
/// With native .NET merging:
/// - feature1 would be overwritten by feature2 from source 2 (index-based merging, e.g. feature_flags:0:id)
/// - feature2: enabled = false (from source 1, index 1)
/// </remarks>
public class ConfigurationFeatureDefinitionProviderOptions
{
    /// <summary>
    /// Controls whether to disable the custom configuration merging logic for Microsoft schema feature flags and fall back to .NET's native configuration merging behavior.
    /// </summary>
    /// <remarks>
    /// This option only affects Microsoft schema feature flags (e.g. feature_management:feature_flags arrays). 
    /// .NET schema feature flags are not affected by this setting.
    /// </remarks>
    public bool DisableCustomConfigurationMerging { get; set; } = true;
}
```

 You can opt-in the novel "concatenation" behavior. The default behavior is the native .net configuration merge behavior.

**Comment by zhiyuanliang-ms** [comment]
> But I am now unclear on the options. Is one variation "old behavior" (native dotnet configuration system default merging) and the other is "current behavior" (the version we complained about)? Or is it "old behavior" and "new behavior" (merge by IDs)?

Hey, @pinkfloydx33 I forgot to update the comment I left in this thread. The discussion in #550 is clearer. The option we are going to offer is 

```C#
public class ConfigurationFeatureDefinitionProviderOptions
{
    bool CustomConfigurationMergingEnabled { get; set; }
}
```

By the default, it is disabled. The "custom merging" behavior is *NOT* merging by `id`. It is the behavior we introduced in v4.2, i.e. concatenate `feature_flags` array and let the last feature flag definition (with the same id) win.

> Is this a major version change? 

It will be minor version change. The discussion is [here](https://github.com/microsoft/FeatureManagement-Dotnet/issues/550#issuecomment-3211411339)

> Only if the default merging strategy changes. I think you can put these changes out in a minor release as long as you default to the native behavior and then you can change the default in a major release if you were so inclined.

Exactly.

**Comment by zhiyuanliang-ms** [comment]
[4.3.0](https://github.com/microsoft/FeatureManagement-Dotnet/releases/tag/4.3.0) is released.
